Updates project_lifecycle.md on WIP branch (#103)
ossf/tac: process/project-lifecycle.md

# Becoming an OpenSSF Project

> The OpenSSF's mission is to inspire and enable the community to secure the open source software we all depend on. OpenSSF top priority is the security of existing open source project however, there are missing pieces in the tooling needed to improve project security. These pieces are the projects OpenSSF aims to steward and support.

Reviewer: ```suggestion
The OpenSSF's mission is to inspire and enable the community to secure the open source software we all depend on. OpenSSF's top priority is the security of existing open source projects however, there are missing pieces in the tooling needed to improve open source project security. These pieces are the projects OpenSSF aims to steward and support.
```

## Project Life Cycle

> New projects to the OpenSSF, and progression through the project lifecycle, are approved by the Technical Advisory Committee (TAC). Projects oversight is provided by either the Technical Advisory Committee (TAC) or a specific Working Group (WG). When a Project reports into a specific WG, that WG can support the Project's progression and provide recommendations to the TAC.

Reviewer: ```suggestion
New projects to the OpenSSF, and progression through the project lifecycle, are approved by the Technical Advisory Committee (TAC). A project's oversight is provided by either the Technical Advisory Committee (TAC) or a specific Working Group (WG). When a Project reports into a specific WG, that WG can support the Project's progression and provide recommendations to the TAC.
```

Projects follow the Sandbox, Incubating, Graduated and Archived lifecycle stages defined below. Projects that seek widespread adoption and end user use are expected to progress through the stages.

<!-- TOC -->

  - [Stages - Definitions & Expectations](#stages---definitions--expectations)
      - [Sandbox](#sandbox)
      - [Incubating](#incubating)
      - [Graduated](#graduated)

>       - [Archiving](#archiving)

Author: Ack, grammatically makes sense, will revert

* Facilitate alignment with existing Working Groups or Projects if this is desired
* Nurture Projects
* Remove possible legal and governance obstacles to adoption and contribution by ensuring all projects adhere to OpenSSF legal, code of conduct, and IP Policy requirements.

#### Project Responsibilities
* Bi-annual updates to the TAC on technical vision and progress on vision
* Maintain a diversified contributor base (i.e. not a single-vendor project)
* Provides project updates to OpenSSF Marketing Committee as requested

#### Project Support
* Receives guidance on technical direction from TAC
* Receives space for Contributor Summits at OpenSSF bi-annual events
* Receives OpenSSF Code of Conduct Committee support
* Reserved space for project updates in OpenSSF newsletters
* May request infrastructure support from the OpenSSF
* Projects, sponsors, and contributors may not promote the project with the OpenSSF logo or be referred to as an "OpenSSF Project" or "OpenSSF $ProjectName." Projects may say they are, "A sandbox project in the OpenSSF" or "An experimental project in the OpenSSF."

#### Sandbox Entry Requirements and Considerations

> * Projects must have a minimum of two maintiners with different company affiliations.

Reviewer 2: nit: typo on `maintainers`

Author: Incorporated

> * Projects must have a minimum of two maintiners with different company affiliations.

Reviewer: ```suggestion
* Projects must have a minimum of two maintainers with different company affiliations.
```

* Projects must be alinged with the OpenSSF mission _and_ either be a novel approach for existing areas, address an unfulfilled need, or be initial code needed for OpenSSF WG work. It is preferred that extensions of existing OpenSSF projects collaborate with the existing project rather than seek a new project.

> * If contributing an existing Project to the OpenSSF, the contribution must undergo license and IP due diligence by the Linux Foundation (LF).

Reviewer 3: +1 to this change.

The statement about "IP due diligence" is merely documenting existing (if unwritten) policy.

#### Submission Forms for Sandbox

1. [Proposal](TBD)
2. [Legal](TBD)

### Incubating

Incubating projects represent maturing but not fully realized projects. Incubating projects have an established technical vision and working project, but are still refining implementation, discovering (and adapting to) new use cases, and operating as an open source team.

#### Project Responsibilities
* Bi-annual updates to the TAC on technical vision and progress on vision
* Maintain a diversified contributor base (i.e. not a single-vendor project) with an active flow of contributions
* Follow security best practices (as recommended by the OpenSSF and others)
* Maintain a point of contact for vulnerability reports

> * Implement, practice, and refine mature software development and release practices such a following a version schema

Reviewer: Entirely a suggestion here -- this new bullet is intended to get projects set up for when they reach that higher level of adoption and maturity.
```suggestion
* Implement, practice, and refine mature software development and release practices such a following a version schema
* Begin to establish the appropriate governance that enable its sustainment for potential graduation
```

Author: +1

Reviewer 4: +1

Reviewer 3: +1

#### Project Support
* Receives guidance on technical direction from TAC
* Receive a TAC or WG sponsor for guidance as an Incubation project
* Receives space for Contributor Summits at OpenSSF bi-annual events
* Receives OpenSSF Code of Conduct Committee support
* Receives infrastructure support (details determined by project leads and OpenSSF Budget Committee)
* Receives support with vulnerability disclosure from the OpenSSF (Vulnerability Disclosure WG)

> * May post project updates and tutorials to the OpenSSF blog

Reviewer 5: @dlorenc 's comment on line #112 applies here as well.

Author: Resolving; will be addressed by Marketing Committee

* May request OpenSSF budget for project improvements such as security audits or time-bound contracting needs
* With additional TAC or WG approval, may fundraise for dedicated project funds, coordinated by the OpenSSF
* Projects may use the OpenSSF logo to promote their project (in accordance with the trademark guidelines). Projects may not be referred to as an "OpenSSF Project" or "OpenSSF $ProjectName." Projects may say they are an "OpenSSF Incubating Project."

#### Incubating Entry Requirements and Considerations

> * Projects must have a minimum of three maintainers with a minimum of two different company affiliations.

Reviewer 5: @inferno-chromium The discussion on Issue #101 has turned into a broader discussion than what was intended. It was really meant to be about the specific requirement here though.

Unfortunately due to the number of changes involved in this PR the shown diff is pretty useless but this requirement used to be "a healthy number of maintainers". This PR has changed it to "a minimum of three maintainers with a minimum of two different company affiliations".

If the TAC is happy with that we can close #101 as far as I'm concerned.

* Projects should be able to show adoption by multiple parties and that adoption's value to the open source community and or end users (may include adoption of beta/early versions).
* Projects must be aligned with the OpenSSF mission and either be a novel approach for existing areas, address an unfulfilled need, or be initial code needed for OpenSSF WG work. It is preferred that extensions of existing OpenSSF projects collaborate with the existing project rather than seek a new project.
* Projects must have documented, initial project governance

#### Project Process: Sandbox to Incubating

> Incubating Projects are required to undergo technical due diligence as a part of the process to move from Sandbox to Incubation. Technical Due Diligence is driven by a TAC or parent WG sponsor. Once the diligence is complete and the proposal made, two weeks are allocated for public comment before a TAC vote is called.

Reviewer: ```suggestion
Incubating Projects are required to undergo technical due diligence as a part of the process to move from Sandbox to Incubation. Technical Due Diligence is driven by a TAC or parent WG sponsor with close collaboration of the project. Once the diligence is complete and the proposal made, the Due Diligence document is made available to the community for two weeks the solicit public comment before a TAC vote is called.
```

> Incubating Projects are required to undergo technical due diligence as a part of the process to move from Sandbox to Incubation. Technical Due Diligence is driven by a TAC or parent WG sponsor. Once the diligence is complete and the proposal made, two weeks are allocated for public comment before a TAC vote is called.

Reviewer 3: As long as there is a consistent location (or channel) for these notices to be sent, I think it is fine. I suggest the TAC's public mailing list for this purpose; incidentally, this is equivalent to how the CNCF process works as well.

#### Submission Forms for Incubating

1. [Proposal](TBD)
2. [Legal](TBD)

### Graduated

Graduated projects signal the highest level of maturity for an OpenSSF project. Graduated projects have consistent, mature open source development practices with sustained contribution and activity. 

#### Project Reponsibilities

* Bi-annual updates to the TAC on technical vision and progress on vision
* Maintain a diversified contributor base (i.e. not a single-vendor project) with an active flow of contributions

> * Follow security best practices (as recommended by the OpenSSF and others)

Reviewer: Is it worth being more explicit here in that we want them to achieve Passing or Silver badging criteria before they can graduate?

Author: Probably depends on if we want to "hardcode" to the CII badging program, or if "as recommended by the OpenSSF" implies that we expect some level of maturity to be demonstrated through OpenSSF programs (which could include CII badges). 

I'm more inclined to the later and to change to the former if the TAC feels like they're receiving graduation submissions that don't follow best practices, and having a certain badge level would have addressed that.

Reviewer 4: The OSSF Badge project is a vetted set of good security practices.  I feel strongly that all foundation projects should "dogfood" our own practices and get a badge to send the message we practice what we preach.

Reviewer: We should be continually evolving our security expectations as a security foundation.  To that end these badges have been out for a while, silver in my mind is a minimum bar.

Author: @AevaOnline @bobcallaway @inferno-chromium @SecurityCRob -- any +1s on Silver?

Reviewer 6: +1 to Silver, esp since it brings in the security testing requirements. Our own security tool need to pass the basic security criteria we preach to others (as @SecurityCRob said)

Reviewer 7: +1 to silver as well.

* Maintain a point of contact for vulnerability reports and follow coordinated vulnerability disclosure practices
* Implement, practice, and refine mature software development and release practices, such as adherence to semantic versioning, and having a declared policy for stable releases and backported fixes.

#### Project Support
* Receives guidance on technical direction from TAC 
* Receives space for Contributor Summits at OpenSSF bi-annual events
* Receives OpenSSF Code of Conduct Committee support
* Receives infrastructure support (details determined by project leads and OpenSSF Budget Committee)
* Receives support with vulnerability disclosure from the OpenSSF (Vulnerability Disclosure WG)

> * May post project updates and tutorials to the OpenSSF blog

Reviewer 2: I think this one will take some negotiation with the GB - they currently control the entire blog process and are likely to continue to want some oversight/direction over the content.

Reviewer: 💯 This may be something that can be generally managed with support from OpenSSF staff.  Regardless we should be clear on what's those areas of responsibility and hand off look like.

Author: Is this a question we can delegate to OpenSSF Marketing Committee? ( cc @jorydotcom) Sounds like the question something like, "What is a blog process that gives access to project maintainers but gives the GB editorial oversight and the TAC awareness?"

#### Graduated Project Entry Requirements and Considerations

* Projects must have maintainers with a minimum of three different company affiliations.
* Projects must be aligned with the OpenSSF mission and be a novel approach for existing areas or address an unfulfilled need. Projects with duplicate, similar or competing fuctionality to an existing OpenSSF project may be denied Graduation status if the TAC does not see technical justification for overlapping projects. 
* Projects must be able to show adoption by multiple parties, which could be production deployments or substantial use by established open source communities, and demonstrate the value of that adoption to either the end users or the open source community.
* Projects must be able to show a consistent release cadence.
* Projects must have documented project governance and be able to demonstrate that governance in action.
* When applicable, projects must have completed a security audit through a third party and addressed audit findings and recommendations

#### Project Graduation Process: Incubating to Graduation

> Graduation requires undergoing due diligence as a part of the process to move from Incubation to Graduation. Due Diligence is driven by a TAC or parent WG sponsor. Once the diligence is complete and the proposal made, two weeks are allocated for public comment before a TAC vote is called.

Reviewer: ```suggestion
Graduation requires undergoing due diligence as a part of the process to move from Incubation to Graduation. Due Diligence is driven by a TAC or parent WG sponsor. For projects seeking Graduation, this may be a light refresh of the existing Due Diligence to cover the additional criteria, or a more in depth Due Diligence depending on the level of change the project has incurred since the original Due Diligence of Incubation was performed. Once the diligence is confirmed by the Sponsor to be complete and the proposal made, the Due Diligence document is made available to the community for two weeks the solicit public comment before a TAC vote is called.
```

Reviewer 3: LGTM

#### Submission Forms for Graduation

1. [Proposal](TBD)
2. [Legal](TBD)

### Archiving

Open source projects have a lifecycle and there are times that projects become inactive due to a variety of reasons. There are also cases where a project may no longer want to be supported by the OpenSSF, or the OpenSSF TAC may no longer wish to recommend the use of a project. Archiving happens through a vote of the TAC, and can be requested by the corresponding project's lead(s) or a TAC member.

#### Archiving Considerations

When voting on a proposal to archive a project, TAC members may consider:
* If the project has consistently met the Project Requirements
* If the project still supports the OpenSSF mission
* If the project's inactivity or inconsistent maintenance presents a user security risk
* If the project design or approach is no longer a recommended best practice

#### Archive Process

Archived projects are no longer in active development and are only archived after a TAC vote.

What does archiving for an OpenSSF project mean?
* OpenSSF will no longer provide budgetary, event, or marketing support for the project

> * Project maintainers may request trademark transfers from the OpenSSF and the Linux Foundation

Reviewer 3: "may request" is, IMHO, appropriately open-ended language, and imposes no obligation upon the LF. 

However, I think the previous line should remain in the doc.

> * Project maintainers may request trademark transfers from the OpenSSF and the Linux Foundation

Reviewer 5: I agree with @AevaOnline though. I think we should have both:
 * Trademarks and domain names of archived projects are still hosted by the OpenSSF and the Linux Foundation 
 * Project maintainers may request trademark transfers from the OpenSSF and the Linux Foundation
So that if no transfer is requested (and approved) the trademarks aren't just dropped.

> * Project maintainers may request trademark transfers from the OpenSSF and the Linux Foundation

Reviewer 7: +1 for option to ask, but no obligation to release the mark.

> * Project maintainers may request trademark transfers from the OpenSSF and the Linux Foundation

Author: @lehors good text clarification -- my edit didn't mean to imply they were auto-released. I'll edit.
